docs/type_inference.md

```diff
@@ -0,0 +1,60 @@
+
+
+## (P3) Merge verifiers into shape functions
+
+Both 
+- **verifiers**: implemented by `Op::verify()`, and 
+- **shape functions**: implemented by `InferTypeOpInterface` like `Op::inferReturnTypes()` or `Op::inferReturnTypeComponents` 
+
+may have verification code to check operands/attributes/results. An ideal split would be that: let the verifiers check the operands/attributes, then let shape functions to calculate inferred result types and check the compatibility against the real result types. However, in reality this split has a few problems:
+
+1. Duplicated code: for example in verifiers we do some processing on the operands then verify some intermediate results, then in shape functions these intermediate results are useful to infer the final results. These intermediate results have to be calculated twice.
+2. Maintenance burden: as verifications of an op are contained in two different method. 
+
+Thus, in most cases the verifiers will be removed and all the verification logic will be merged into the shape functions (example: [PR#135](https://github.com/openxla/stablehlo/pull/135)). The only exceptions will be:
+
+1. Op that cannot have a shape function: like `stablehlo.reshape`, whose result type cannot be inferred.
+2. (should be rare) Op that have tons of logic in their verifier but very little logic in their shape function.
+
+
+## (P4) Align error messages by following standard templates. 
+
+Different error messages for similar types of error have inconsistent user experience and more maintenance efforts. We propose using utils to wrap each type of common errors and apply gradually to the existing code.
+
+
+## (P5) Establish testing guidelines
+
+**Do we need to add/maintain tests for verifications that are covered by ODS?**
+
+They are expected to be optional. The tests should focus on the verifiers and shape functions, while changes to ODS need a revisit of this op.
+
+But stay careful about the missing pieces: for example, if the op contains `SameOperandsAndResultShape` in its definition, the element types of operands/results still need to be checked with tests. 
+
+**Where do we put tests for verifiers and type inference?**
+
+Verifications are kept in [ops\_stablehlo.mlir](https://github.com/openxla/stablehlo/blob/main/stablehlo/tests/ops_stablehlo.mlir) and [infer\_stablehlo.mlir](https://github.com/openxla/stablehlo/blob/main/stablehlo/tests/infer_stablehlo.mlir). If an op is complicated and could contain a lot of tests, consider adding a separate test file named `verify_<op_name>.mlir` or` verify_<your_topic>.mlir` within the same folder.
+
+
+## (P6) Improve our infrastructure to better support dynamism
+
+StableHLO supports dynamism, and has experimental support for bounded dynamism. In perspective of type inference, certain improvements and modifications are needed. Please refer to [bounded_dynamism.md](https://github.com/openxla/stablehlo/blob/main/stablehlo/docs/bounded_dynamism.md) for detailed plans. 
```

Comment:
Given that we don't have this document yet, let's drop this section.

Reply:
Yes.